### ISSUE TITLE
GlideN64 Fullscreen showing black borders?

### ISSUE THREAD

**Issue by naddel81** [gonetz/GLideN64#1457]
Hi,

I really love GlideN64 for its remarkable 16:9 feature. It works amazingly well in some games with minor pop-ins. But regardless of which mode I use, there are always those annoying black borders:

![img_5830](https://cloud.githubusercontent.com/assets/16635406/24675753/a34d7a4e-1980-11e7-85b2-adbebaabd78a.JPG)

Those do not show in Glide64 or Jabos Plugin. How can I make GlideN64 stretch out to the full screen and use all its potential?

Best wishes!

**Comment by loganmc10** [comment]
Have you tried setting "Crop Mode" to "Auto"? I don't know if it is compatible with the widescreen hack, but that normally gets rid of black bars.

**Comment by oddMLan** [comment]
Some black bars cannot be removed due to VI emulation, which has become an integral part of the plugin. I dunno if for example we can take a picture of the final result and apply cropping to the image, without disabling VI emulation.

**Comment by naddel81** [comment]
Thanks for your quick replies! I tried auto-cropping, but it does not change anything.
So there is no way to implement a "auto-fullscreen" option as in Glide64 or Jabos Plugin?

By the way: How come the 16:9 feature is possible without stretching? Did the N64 render the image in widescreen even in games that did not support that feature (like Mario 64)?

**Comment by loganmc10** [comment]
> By the way: How come the 16:9 feature is possible without stretching? Did the N64 render the image in widescreen even in games that did not support that feature (like Mario 64)?

No, it's just a hack, you can read a little bit about it here:
http://gliden64.blogspot.com/2014/12/widescreen-hack_1.html

**Comment by naddel81** [comment]
thanks for the link. the last thing to make GlideN64 my favourite GFX plugin would be the "fit to screen" option or some adjustments that we can do ourselves to stretch the image properly and ban the black borders.

**Comment by retrobenny** [comment]
If it had zoom options of its own,you could do like I can on Android like how I use 105% zoom to make it mostly fit the screen without cut-off.

Also,do you have my culling fix code for Banjo-Kazooie?
https://banjosbackpack.com/forums/showthread.php?8323-B-K-No-Culling-GS-Code-(for-widescreen)

**Comment by naddel81** [comment]
a custom zoom option would be great. can this be implemented? also what does culling mean?

best wishes!

**Comment by gonetz** [comment]
> Did the N64 render the image in widescreen even in games that did not support that feature 

3D games usually render more graphics than displayed to user. Part of scene, which do not fit the viewport is cut. Widescreen hack extends vewport to show you more. This is not always work because rendering out of default viewport is waste of console computational resources and game developers are trying to cull rendering of currently unseen objects.

Black boarders, which can't be remove by crop (do not rely on auto mode btw, it not always works) added by Video Interface. Of course, VI emulation also can be hacked to remove black boarders. As any hack, it will help to some games and break others. I don't want to hack code which I just made working properly. May be later.

**Comment by AmbientMalice** [comment]
This is a distinctly "that's how it's supposed to be" issue. A real N64 behaves exactly the same way. The borders are a deliberate choice by the developers.

**Comment by naddel81** [comment]
thanks to all for letting me know. I was just wondering because Glide64 does fullscreen without the typical black borders (even on games that had the borders back the ). how does glide64 do it? is it using another (superior) rendering method? I do not want to offend anybody, I am just curious. thanks in advance!

**Comment by gonetz** [comment]
> is it using another (superior) rendering method?

Rather opposite. It emulates Video Interface on very primitive level and does not take into account offsets set by VI.

**Comment by naddel81** [comment]
so in other words I have to live with the borders when I want full effects (frame buffer emulation) and best compatibility + 16:9 widehack, right?

**Comment by gonetz** [comment]
Sure. If game disigned to have black borders, good emulator will show them to you.

**Comment by oddMLan** [comment]
Pardon my ignorance but, can't the final image with borders added by the VI be passed to another buffer, and then do necessary postprocessing to crop the VI borders to taste? That way in theory you would have the accuracy of VI emulation, without the borders. Or it's not that easy?

Or would it be easier just to disable VI emulation altogether?

**Comment by AmbientMalice** [comment]
I'm not sure why some people are so allergic to black boarders, to be honest. The game was created to be viewed that way. You're already playing 4:3 games on a 16:9, 16:10, or 2.35:1 ratio monitor with massive black bars on either side. What difference does relatively thin black borders make?

Cropping them in post is an interesting idea, that said.

**Comment by gonetz** [comment]
Crop as post filter is not hard to implement. If you select internal rendering resolution high enough (large scale of native resolution), you probably will get no decrease of visual quality. 

If you want to disable VI emulation altogether, just disable frame buffer emulation.

**Comment by oddMLan** [comment]
>The game was created to be viewed that way. 

Not necessarily. Games that had those black borders were taking in account the overscan of old CRT TVs. So games actually were designed to have enough black areas surrounding the picture so presentation wasn't affected, and effectively showing a borderless picture to the player.
Think of this postfilter as an effective "overscan" emulation. However I don't think it would work automatically. You could probably set it three ways. A "safe" value that could be averaged from some of the games' existing black borders. An "auto" option probably would be too hacky to work around, because it would crop anything black, so transitions wouldn't look right (unless it could be guessed from the game's VI parameters... how does the current "auto" cropping method works by the way?). And a custom "zoom" option, much like good ol DVD players (and some modern TVs) allow you to set. Or perhaps just allow the user set how many pixels to crop from each side, if they don't care about preserving the original aspect ratio.

**Comment by AmbientMalice** [comment]
@oddMLan IMO, it's a mixture of the two. Conker's black borders are because the game uses a slightly lower than normal resolution, for example. I feel the key problem here is that you can't really reproduce CRT scaling and overscan behavior on a computer monitor because you get unpleasant blurring.

**Comment by oddMLan** [comment]
>you can't really reproduce CRT scaling and overscan behavior on a computer monitor because you get unpleasant blurring

Not if you set it higher than your current monitor resolution, which can be achieved with 4x+ Native Resolution factor.

**Comment by dankcushions** [comment]
seems like conker had a [partially visible border even on native hardware + CRT](https://www.reddit.com/r/n64/comments/4w1q62/conkers_bad_fur_day_resolution/d63b43v/), and a [full border on native hardware + LCD TV](https://www.reddit.com/r/n64/comments/4w1q62/conkers_bad_fur_day_resolution/).

not sure about other games, but as PAL gamer i'm very used to seeing borders on actual hardware + CRT :)

**Comment by AmbientMalice** [comment]
>not sure about other games, but as PAL gamer i'm very used to seeing borders on actual hardware + CRT :)

This.

**Comment by naddel81** [comment]
> Crop as post filter is not hard to implement. If you select internal rendering resolution high enough (large scale of native resolution), you probably will get no decrease of visual quality.

would be awesome because as seen on the picture above there is 15-20% blackness all around the image. glideN64 would really combine the best of both worlds when we could just stretch that image out to our liking.

**Comment by ulfman** [comment]
Hello :)

Crop the Image to fill the entire Screen is working nice with US (NTSC) Games. But what is with EUROPE (PAL) Games? Doesnt work at all. Can you please fix this?
